Improvements to particle entity.

hyperlogic opened the pull request:
* Changed particle geometry to billboarded quads
* Added texture support
* Added ajt-test.js particle test script.
* GeometryCache support for batched quads with texCoords.
* Bug fix for infinite loop if _lifetime was very large.
* Don't reset the simulation on animation loop.
* stop emitting particles on animation stop, but keep simulating until there are no more living particles.
* Removed some trailing whitespace

hifi-gustavo commented:
An admin must verify this patch for the pull request build to fire. Type "ok to test" to begin testing this pull request or "add to whitelist" to test this and all future pull requests from this collaborator.

PhilipRosedale commented:
ok to test

hifi-gustavo commented:
Build console output for commit: ede42285b1ba1c6d1d2045c0be42ab699e684ab9 available at: https://gist.github.com/6cdabc5e3765ec264d10<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4820.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4820.dmg

ZappoMan commented on libraries/entities-renderer/src/RenderableParticleEffectEntityItem.cpp:
```diff
@@ -29,62 +30,155 @@ RenderableParticleEffectEntityItem::RenderableParticleEffectEntityItem(const Ent
+    PerformanceTimer perfTimer("RenderableParticleEffectEntityItem::render");
+
+    if (_texturesChangedFlag)
+    {
```
coding standard: brace placement... should be:

```
if (_texturesChangedFlag)  {
```

ZappoMan commented on libraries/entities-renderer/src/RenderableParticleEffectEntityItem.cpp:
```diff
@@ -29,62 +30,155 @@ RenderableParticleEffectEntityItem::RenderableParticleEffectEntityItem(const Ent
-    // Right now we're just iterating over particles and rendering as a cross of four quads.
-    // This is pretty dumb, it was quick enough to code up.  Really, there should be many
-    // rendering modes, including the all-important textured billboards.
+    glm::vec3 up_offset = args->_viewFrustum->getUp() * pa_rad;
```
coding standard: variable names should use camel case

```
glm::vec3 upOffset = args->_viewFrustum->getUp() * particleRadius;
```

ZappoMan commented on libraries/entities-renderer/src/RenderableParticleEffectEntityItem.cpp:
```diff
@@ -29,62 +30,155 @@ RenderableParticleEffectEntityItem::RenderableParticleEffectEntityItem(const Ent
+
+void RenderableParticleEffectEntityItem::renderUntexturedQuads(RenderArgs* args) {
+
     float pa_rad = getParticleRadius();
```
coding standard: variable names should use camel case

```
float particleRadius = getParticleRadius();
```

ZappoMan commented on libraries/entities-renderer/src/RenderableParticleEffectEntityItem.cpp:
```diff
@@ -29,62 +30,155 @@ RenderableParticleEffectEntityItem::RenderableParticleEffectEntityItem(const Ent
+    glm::vec3 right_offset = args->_viewFrustum->getRight() * pa_rad;
 
     QVector<glm::vec3>* pointVec = new QVector<glm::vec3>(_paCount * VERTS_PER_PARTICLE);
+    quint32 paCount = 0;
```
coding standard: variable names should be fully descriptive

```
quint32 particleCount = 0;
```

ZappoMan commented on libraries/entities-renderer/src/RenderableParticleEffectEntityItem.cpp:
```diff
@@ -29,62 +30,155 @@ RenderableParticleEffectEntityItem::RenderableParticleEffectEntityItem(const Ent
-    // This is pretty dumb, it was quick enough to code up.  Really, there should be many
-    // rendering modes, including the all-important textured billboards.
+    glm::vec3 up_offset = args->_viewFrustum->getUp() * pa_rad;
+    glm::vec3 right_offset = args->_viewFrustum->getRight() * pa_rad;
```
coding standard: variable names should use camel case

```
glm::vec3 rightOffset = args->_viewFrustum-> getRight() * particleRadius;
```

ZappoMan commented on libraries/entities-renderer/src/RenderableParticleEffectEntityItem.cpp:
```diff
@@ -29,62 +30,155 @@ RenderableParticleEffectEntityItem::RenderableParticleEffectEntityItem(const Ent
+    }
+
+    // sort particles back to front
+    qSort(posVec->begin(), posVec->end(), zSort);
```
consider using lambda function for this

ZappoMan commented on libraries/entities-renderer/src/RenderableParticleEffectEntityItem.cpp:
```diff
@@ -29,62 +30,155 @@ RenderableParticleEffectEntityItem::RenderableParticleEffectEntityItem(const Ent
+
+void RenderableParticleEffectEntityItem::renderTexturedQuads(RenderArgs* args) {
+
+    float pa_rad = getParticleRadius();
```
coding standard: variable names

ZappoMan commented on libraries/entities-renderer/src/RenderableParticleEffectEntityItem.cpp:
```diff
@@ -29,62 +30,155 @@ RenderableParticleEffectEntityItem::RenderableParticleEffectEntityItem(const Ent
+    float pa_rad = getParticleRadius();
+
+    const float MAX_COLOR = 255.0f;
+    glm::vec4 paColor(getColor()[RED_INDEX] / MAX_COLOR, getColor()[GREEN_INDEX] / MAX_COLOR,
```
coding standard: variable names

ZappoMan commented on libraries/entities-renderer/src/RenderableParticleEffectEntityItem.cpp:
```diff
@@ -29,62 +30,155 @@ RenderableParticleEffectEntityItem::RenderableParticleEffectEntityItem(const Ent
+    // sort particles back to front
+    qSort(posVec->begin(), posVec->end(), zSort);
+
+    QVector<glm::vec3>* vertVec = new QVector<glm::vec3>(_paCount * VERTS_PER_PARTICLE);
```
coding standard: variable names - maybe use "vertices"... we don't need to know that it's a vector

ZappoMan replied:
why did you make this heap allocated?

ZappoMan commented on libraries/entities-renderer/src/RenderableParticleEffectEntityItem.cpp:
```diff
@@ -29,62 +30,155 @@ RenderableParticleEffectEntityItem::RenderableParticleEffectEntityItem(const Ent
+    qSort(posVec->begin(), posVec->end(), zSort);
+
+    QVector<glm::vec3>* vertVec = new QVector<glm::vec3>(_paCount * VERTS_PER_PARTICLE);
+    QVector<glm::vec2>* texCoordVec = new QVector<glm::vec2>(_paCount * VERTS_PER_PARTICLE);
```
coding standard: variable names - maybe use "textureCoords"... we don't need to know that it's a vector

ZappoMan commented on libraries/entities-renderer/src/RenderableParticleEffectEntityItem.cpp:
```diff
@@ -29,62 +30,155 @@ RenderableParticleEffectEntityItem::RenderableParticleEffectEntityItem(const Ent
+    QVector<glm::vec3>* vertVec = new QVector<glm::vec3>(_paCount * VERTS_PER_PARTICLE);
+    QVector<glm::vec2>* texCoordVec = new QVector<glm::vec2>(_paCount * VERTS_PER_PARTICLE);
+
+    s_dir = args->_viewFrustum->getDirection();
```
coding standard: variable names

ZappoMan commented on libraries/entities-renderer/src/RenderableParticleEffectEntityItem.cpp:
```diff
@@ -29,62 +30,155 @@ RenderableParticleEffectEntityItem::RenderableParticleEffectEntityItem(const Ent
+    QVector<glm::vec2>* texCoordVec = new QVector<glm::vec2>(_paCount * VERTS_PER_PARTICLE);
+
+    s_dir = args->_viewFrustum->getDirection();
+    glm::vec3 up_offset = args->_viewFrustum->getUp() * pa_rad;
```
coding standard: variable names

ZappoMan commented on libraries/entities-renderer/src/RenderableParticleEffectEntityItem.cpp:
```diff
@@ -29,62 +30,155 @@ RenderableParticleEffectEntityItem::RenderableParticleEffectEntityItem(const Ent
+
+    s_dir = args->_viewFrustum->getDirection();
+    glm::vec3 up_offset = args->_viewFrustum->getUp() * pa_rad;
+    glm::vec3 right_offset = args->_viewFrustum->getRight() * pa_rad;
```
coding standard: variable names

ZappoMan commented on libraries/entities-renderer/src/RenderableParticleEffectEntityItem.cpp:
```diff
@@ -29,62 +30,155 @@ RenderableParticleEffectEntityItem::RenderableParticleEffectEntityItem(const Ent
+    glm::vec4 paColor(getColor()[RED_INDEX] / MAX_COLOR, getColor()[GREEN_INDEX] / MAX_COLOR,
+                      getColor()[BLUE_INDEX] / MAX_COLOR, getLocalRenderAlpha());
+
+    QVector<glm::vec3>* posVec = new QVector<glm::vec3>(_paCount);
```
why did you make this heap allocated?

ZappoMan commented:
Couple of general comments:
* I realize that I left some Coding Standard notes on code that wasn't new to this PR, we shouldn't have let the original code get past a code review as it didn't meet the standard. To try to keep the code consistent, we will go back and repair coding standard issues whenever we touch a file... so if you could repair those variable names too, that is appreciated.

* You need to also bump the version number of the entity data packets. Because you don't want new versions of the server to start sending texture data to old versions of the client that don't know about it. To do this you need to do the following:

**In libraries/networking/src/PacketHeaders.h around line 145**
```
const PacketVersion VERSION_ENTITIES_ZONE_ENTITIES_HAVE_ATMOSPHERE = 20;
const PacketVersion VERSION_ENTITIES_ZONE_ENTITIES_HAVE_SKYBOX = 21;
const PacketVersion VERSION_ENTITIES_ZONE_ENTITIES_STAGE_HAS_AUTOMATIC_HOURDAY = 22;
const PacketVersion VERSION_ENTITIES_PARTICLE_ENTITIES_HAS_TEXTURES = 23;
```

**In libraries/networking/src/PacketHeaders.cpp around line 75**
```
        case PacketTypeEntityAddOrEdit:
        case PacketTypeEntityData:
            return VERSION_ENTITIES_PARTICLE_ENTITIES_HAS_TEXTURES;
```

hyperlogic commented:
Thanks for the comments. Yeah, I can go back and fix-up the existing variable names to follow the coding standards. Initially, I just wanted to make a minimal diff, to make review easier.

About the heap allocations, so yeah, it's not great to be making allocations per particle entity during the rendering traversal. What I can do instead is create class static vectors that are re-used for every particle entity, eventually those vectors will hit a high water mark and never need to be allocated again.  

While i'm at it, I think I'd also like to make the member variables that hold the particle data also be vectors. I noticed there's a pretty bad bug if you dynamically change the max number of particles after the particle entity has been created.

Just a question, it looks like you prefer Qt library functions over std. i.e. QVectors over std::vector, QSharedPointer over std::shared_ptr, should I continue with that pattern?

ZappoMan commented:
Use Qt classes for containers. They are used throughout the code, and have many preferable advantages like copy on write and other optimizations.

ZappoMan commented:
re: "About the heap allocations, so yeah, it's not great to be making allocations per particle entity during the rendering traversal."

My point was less about the issue of allocating during rendering (ultimately, you need to change the particle positions, so the data in the gpu::buffer needs to change)...

My point was just that you could have had a QVector<gml::vec3> as a local variable in the function and your code would have been almost similar, but would have removed the need to do a delete at the bottom of the function... I wasn't speaking of the vector data (which will obviously be heap allocated)... I was speaking of the vector itself, which can be a local variable.

hyperlogic commented:
re: "My point was just that you could have had a QVector&lt;gml::vec3&gt;", understood, this is much cleaner.

hyperlogic commented:
Commit b3af515 should address your concerns w.r.t. the Coding Standards.  While going over that code I also fixed some issues I discovered...

  * Updated variable naming to match coding standards.
  * Changed particle raw float arrays to vectors.
  * Bug fix: changing maxParticles will no longer lead to memory corruption.
  * Made particle ring buffer more explicit, added _particleTailIndex.
  * Added getLivingParticleCount() private method.
  * Moved integration and bounds code into private methods.
  * Bug fix: high emit rates now properly integrate particles forward for the remaing frame time, not the entire frame.
  * Bug fix: new particles were not initiaized to origin.
  * Added more particles to ajt-test.js.
  * Bug fix: ajt-test.js script was not shutting down properly.
  * Removed random seed, unless we have a psudo random number generator per particle system instance, it's unlikely that this will preserve sync.
  * Bumped packet version number.

Because this branch was a bit out of date, I went ahead and merged highfidelity:master into it.  Let me know if you'd prefer a rebase.

hifi-gustavo commented:
Build console output for commit: b3af5152240c20583f93855906b0beac66d50c35 available at: https://gist.github.com/6cdabc5e3765ec264d10<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4820.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4820.dmg

hifi-gustavo commented:
Build console output for commit: e1cbfcd901c6c88c6c9faab62af10ebcf640115c available at: https://gist.github.com/6cdabc5e3765ec264d10<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4820.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4820.dmg
